returnn/datasets/postprocessing.py

```diff
@@ -0,0 +1,253 @@
+            :return: generator providing data samples in the form of a TensorDict
+            """
+            data_keys = self._dataset.get_data_keys()
+
+            seq_index = 0
+            while self._dataset.is_less_than_num_seqs(seq_index):
+                self._dataset.load_seqs(seq_index, seq_index + 1)
+                data = {data_key: self._dataset.get_data(seq_index, data_key) for data_key in data_keys}
+                data["seq_tag"] = str_to_numpy_array(self._dataset.get_tag(seq_index))
+                tensor_dict = self._data_dict_to_tensor_dict(data)
+                if self._map_seq is not None:
+                    tensor_dict = self._map_seq(tensor_dict)
+                    assert isinstance(tensor_dict, TensorDict), "map_seq must produce a TensorDict"
+                yield tensor_dict
+                seq_index += 1
+
+        data_iter = _iterate_dataset()
+        if self._map_seq_stream is None:
+            return data_iter
+
+        data_iter = self._map_seq_stream(data_iter)
+        assert isinstance(data_iter, Iterator), "map_seq_stream must produce an Iterator"
+        return data_iter
+
+    def _data_dict_to_tensor_dict(self, data_dict: Dict[str, ndarray]) -> TensorDict:
+        """
+        :return: the given data dict converted to a TensorDict class
+        """
+
+        def _make_tensor(name: str, data: ndarray) -> Tensor:
+            sparse_dim = None
+            if data.dtype.name.startswith("str"):
+                dims = []
+                dtype = "string"
+            else:
+                dims, sparse_dim = self._dim_cache.get(name, (None, None))
+                dtype = data.dtype
+
+            if dims is None:
+                feature_dims = [
+                    Dim(dimension=v, name=f"{name}_dim{i + 1}")
+                    for i, v in enumerate(self._dataset.get_data_shape(name))
+                ]
+                dims = [Dim(dimension=None, name=f"{name}_frame"), *feature_dims]
+                if self._dataset.is_data_sparse(name):
+                    sparse_dim = Dim(
+                        dimension=self._dataset.get_data_dim(name) if self._dataset.is_data_sparse(name) else None,
+                        name=f"{name}_sparse",
+                    )
+                    if name in self._dataset.labels:
+                        sparse_dim.vocab = Vocabulary.create_vocab_from_labels(self._dataset.labels[name])
+                self._dim_cache[name] = (dims, sparse_dim)
+
+            try:
+                return Tensor(name, dims=dims, dtype=dtype, sparse_dim=sparse_dim, raw_tensor=data)
+            except ShapeMismatchException as exc:
+                raise ShapeMismatchException(
+                    "Could not convert from mapping function output to `TensorDict`, "
+                    f"do the data shapes {data.shape} match up with the ones declared "
+                    f"in `map_output` (inferred to: {dims}, sparse={sparse_dim})?"
+                ) from exc
```

Comment:
```suggestion
            return Tensor(name, dims=dims, dtype=dtype, sparse_dim=sparse_dim, raw_tensor=data)
```

Reply:
There should never be an exception here, so there is no point in catching something which can never happen.